### ISSUE TITLE
Signal Relation Incosistency

### ISSUE THREAD

**Issue by EMREOYUN** [GNimrodG/simrail-map-optimized#23]
When used locally, some signals connect to the wrong signals in the relation. Does the relations required to be fixed my hand?

Wrongly connected relationship:
![opera_T062LeHaeS](https://github.com/GNimrodG/simrail-map-optimized/assets/35144629/bc5bc464-9a6c-4010-8113-c785672c2583)

The correct one:
![image](https://github.com/GNimrodG/simrail-map-optimized/assets/35144629/b0eba508-3267-44fc-bc16-42451e5036e2)

**Comment by GNimrodG** [comment]
Yeah it's known bug... sometimes the trains just teleport or something happens to them. 
I haven't found a solution to this yet, other than manually checking the signals and removing the broken connections.
I think the problem is the TTL [here](https://github.com/GNimrodG/simrail-map-optimized/blob/e5f6e534ed725e21a3f60bb43dec25b7b25329ed/smo-backend/src/analytics/signal-worker.ts#L18C1-L18C2), I tried lowering it from the previous 1 hour to 5 min that didn't help much...

If I connect to the DB with QGIS to visualize the connections there are lots of broken lines there:
![image](https://github.com/GNimrodG/simrail-map-optimized/assets/25616517/77e3eec1-ae09-4070-a528-272b9107c8e6)

**Comment by GNimrodG** [comment]
Block signals have a rule to them so they don't get that messed up but for main signals I couldn't find a solution yet.

**Comment by EMREOYUN** [comment]
It seems some commits made the signal connections are working correctly. I've been running the server for a while with a fresh instance and signals getting matched with correct signals.

**Comment by GNimrodG** [comment]
Yeah I limited the TTL of trains' old signal information, so it expires when a server restarts. There are still occasionally some wrong connections though.

**Comment by EMREOYUN** [comment]
Which ones? I can look it up on my server to confirm.

**Comment by GNimrodG** [comment]
DZ_G11 connected to DZ_G13 which shouldn't happen because they're parallel to each other. And even DZ_G13 connected to DZ_G11 so they're each others next AND prev signals. Same with Dra_A and Dra_E.
I actually wrote a query to detect this:
```sql
SELECT signal_connections.prev,
       signal_connections.next,
       signal_connections.created_at                                   AS first_created_at,
       s.created_at                                                    AS second_created_at,
       concat(signal_connections.prev, '<->', signal_connections.next) AS id
FROM signal_connections
         JOIN signal_connections s
              ON s.prev = signal_connections.next AND s.next = signal_connections.prev
```

Then DW_P (DGW) connected to Bo_D1 (Bukowno) which shouldn't be possible because there is Sławków between them.
Also DS_G connected to DW_P which is wrong because there is DW_B the entry signal to DGW between them.
Also somehow Ju_G (Juliusz) connected to DZ_L1 (DGZ) which is an impossible connection.

**Comment by GNimrodG** [comment]
I also fixed lots of this issues manually so my instance is pretty okay now.
![image](https://github.com/GNimrodG/simrail-map-optimized/assets/25616517/2b82861d-f95e-403f-acae-45a358678c06)

**Comment by EMREOYUN** [comment]
DZ_G11 connected to DZ_F in my server which shouldn't be possible for the same reason you wrote. Dra_A and Dra_E are also connected. I guess I will use [SRTD](https://tplmilano.altervista.org/E/SRTD/#en1_0_0_0) to connect all signals manually with a script since [the source code of SRTD](https://tplmilano.altervista.org/E/SRTD/c.js?v=240220) contains all signals and their connections except before Grodzisk Mazowiecki.

**Comment by GNimrodG** [comment]
I use [QGIS](https://www.qgis.org/en/site/) to visualize the connections. I connected it to the PostGIS DB and created a view which creates a LineString from the `signal_connections`
Here is the query for that:
```sql
SELECT concat(
               CASE
                   WHEN p.next_finalized = true THEN concat('[', p.name, ']')
                   ELSE p.name
                   END, '->',
               CASE
                   WHEN n.prev_finalized = true THEN concat('[', n.name, ']')
                   ELSE n.name
                   END)              AS id, -- without an ID QGIS can't read the data
       signal_connections.prev,
       signal_connections.next,
       signal_connections.distance,
       signal_connections.created_at,
       st_makeline(p.point, n.point) AS geom,
       p.next_finalized              AS start_finalized,
       n.prev_finalized              AS end_finalized,
       p.next_regex                  AS start_regex,
       n.prev_regex                  AS end_regex
FROM signal_connections
         JOIN signals p ON signal_connections.prev = p.name
         JOIN signals n ON signal_connections.next = n.name
```
Then you can just import this view as a layer in QGIS.

I used screenshots of the dispatcher tables to correct the signals on my server manually with hand. I might look into importing connections from SRTD too.

**Comment by GNimrodG** [comment]
Also an intersting thing I noticed that there is a third (I think non-electrified) track between SG and Bedzin, between signal SG_C (which I don't have on my server) and B_P (which has been used enough times to appear on my server) that isn't there on the SRTD. 
SG side:
![image](https://github.com/GNimrodG/simrail-map-optimized/assets/25616517/b6531722-0aa1-44c6-8f92-0257150145ea)
Bedzin side:
![image](https://github.com/GNimrodG/simrail-map-optimized/assets/25616517/0f726779-d86f-4b79-8e3c-3af030fa47a6)

Also on my server there is signal LA_F110 which shouldn't be accessible in the current state of the game.
![image](https://github.com/GNimrodG/simrail-map-optimized/assets/25616517/4d5b24ce-06ad-4b69-8c0f-9d16e5c796c3)

**Comment by EMREOYUN** [comment]
> Also an intersting thing I noticed that there is a third (I think non-electrified) track between SG and Bedzin, between signal SG_C (which I don't have on my server) and B_P (which has been used enough times to appear on my server, discovered at 2024-06-16 19:10:43) that isn't there on the SRTD. 
> SG side:
> ![image](https://github.com/GNimrodG/simrail-map-optimized/assets/25616517/b6531722-0aa1-44c6-8f92-0257150145ea)
> Bedzin side:
> ![image](https://github.com/GNimrodG/simrail-map-optimized/assets/25616517/0f726779-d86f-4b79-8e3c-3af030fa47a6)
> 
> Also on my server there is signal LA_F110 which shouldn't be accessible in the current state of the game.
> ![image](https://github.com/GNimrodG/simrail-map-optimized/assets/25616517/4d5b24ce-06ad-4b69-8c0f-9d16e5c796c3)
> 

Probably those are came from EU3, where there is no rules. I once sent a train to LA_F4M once to test a theory. As far as I know, SRTD only contains accessable routes and might not include every route in the game like Grodzisk Mazowiecki - Warsawa is completely missing and some minor issues that makes trains disappear on the routes. Maybe hopping and sending a train on track 3 from Bedzin might add SG_C on the map as well.

**Comment by GNimrodG** [comment]
So I added a field to log which train created which relation and discovered that errors don't only come from EU3, actually most of them came from PL2, but I've got invalid connections from all the servers at least once, and the creation date doesn't line up with the restarts either... I just don't know what's going on, it's probably sever lag or something that's causing these issues.
